app/src/androidTest/java/ch/epfl/sdp/musiconnect/CloudStorageTest.java

```diff
@@ -0,0 +1,110 @@
+package ch.epfl.sdp.musiconnect;
+
+import android.Manifest;
+import android.content.Context;
+import android.net.Uri;
+
+import androidx.test.platform.app.InstrumentationRegistry;
+import androidx.test.rule.ActivityTestRule;
+import androidx.test.rule.GrantPermissionRule;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+
+import java.util.concurrent.TimeUnit;
+
+import ch.epfl.sdp.BuildConfig;
+import ch.epfl.sdp.R;
+
+import static androidx.test.espresso.Espresso.onView;
+import static androidx.test.espresso.assertion.ViewAssertions.matches;
+import static androidx.test.espresso.matcher.RootMatchers.withDecorView;
+import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static androidx.test.espresso.matcher.ViewMatchers.withText;
+import static org.hamcrest.core.IsNot.not;
+
+public class CloudStorageTest {
+    @Rule
+    public final ActivityTestRule<StartPage> startPageRule =
+            new ActivityTestRule<>(StartPage.class);
+
+    @Before
+    public void stallBefore() {
+        waitALittle(5);
+    }
+
+    @After
+    public void stallAfter() {
+        waitALittle(5);
+    }
+
+    private Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
+    private CloudStorage storage = new CloudStorage(context);
```

Comment:
This is a REALLY bad idea. You must not test the production DB (even not Firebase at all). Here your test uploads and downloads stuff on external service. Tests must not have this kind of behavior, they should only depend on local environment. 
You don't need to test that firebase works as intended.

Here you should do dependency injection:
the *StorageReference* instance in CloudStorage should be taken as argument by the constructor so that you can pass a mocked instance in the test and a real Firebase one in the production code.
@Gondolav

Reply:
Thank you for the review ! Okay, the issue is that I'm not familiar with creating mocks of such services or even creating mocks at all. Is there a tutorial or some guidelines I could follow ? Actually it would solve a lot of testing issues that are popping out now: those tests seem to randomly fail since they rely on the performance of the cloud storage service.

Reply:
Hey,

The main idea between mocking in the way you need to implement it here (without libraries) is the **Dependency Injection**. In a nutshell, the idea is to delegate the choose of a dependency (here Firebase or any other DB services) to the caller of a class instead of setting it directly into the class: 
https://www.freecodecamp.org/news/a-quick-intro-to-dependency-injection-what-it-is-and-when-to-use-it-7578c84fa88f/

This let the caller "inject" the dependency and so, in the prod code you inject Firebase, and in the tests, you inject a class that you wrote that fakes Firebase (has the same interface but does almost nothing but has a behavior you choose to test). Here, as Firebase's interface is a bit complicated to implement, you should encapsulate it into a class you write that will be you DB all around the app (with a implementation using Firebase and one mocked).
If you need more precision, please let me know, I can explain you more precisely. We could even schedule a Zoom meeting if you need.

You can also have a look at libraries that help you mocking part of your code (as Mockito).
For this part, I would recommend the exercises about mocking for SwEng class last semester:
https://github.com/sweng-epfl/public/blob/master/exercises/testing/mocking.md

Hope this helps and don't hesitate to ask if you need more explanation :)